stdenv: let overrideAttrs accept attrset OR function (#239005)
NixOS/nixpkgs: doc/using/overrides.chapter.md

The function `overrideAttrs` allows overriding the attribute set passed to a `stdenv.mkDerivation` call, producing a new derivation based on the original one. This function is available on all derivations produced by the `stdenv.mkDerivation` function, which is most packages in the nixpkgs expression `pkgs`.

Example usages:

```nix
helloBar = pkgs.hello.overrideAttrs (finalAttrs: previousAttrs: {
  pname = previousAttrs.pname + "-bar";
});
```

In the above example, "-bar" is appended to the pname attribute, while all other attributes will be retained from the original `hello` package.

The argument `previousAttrs` is conventionally used to refer to the attr set originally passed to `stdenv.mkDerivation`.

The argument `finalAttrs` refers to the final attributes passed to `mkDerivation`, plus the `finalPackage` attribute which is equal to the result of `mkDerivation` or subsequent `overrideAttrs` calls.

If only a one-argument function is written, the argument has the meaning of `previousAttrs`.

```nix
helloWithDebug = pkgs.hello.overrideAttrs {
  separateDebugInfo = true;
};
```

> In the above example, the `separateDebugInfo` attribute is overridden to be true, thus building debug info for `helloWithDebug`. Function arguments can be omitted if there is no need to access previousAttrs or finalAttrs.

Author: doc added

> ```nix
> helloWithDebug = pkgs.hello.overrideAttrs {
>   separateDebugInfo = true;
> };
> ```
> 
> In the above example, the `separateDebugInfo` attribute is overridden to be true, thus building debug info for `helloWithDebug`. Function arguments can be omitted if there is no need to access previousAttrs or finalAttrs.

Reviewer: ````suggestion
Function arguments can be omitted entirely if there is no need to access `previousAttrs` or `finalAttrs`.

```nix
helloWithDebug = pkgs.hello.overrideAttrs {
  separateDebugInfo = true;
};
```

In the above example, the `separateDebugInfo` attribute is overridden to be true, thus building debug info for `helloWithDebug`.
````

This relates it to the text above.
